File: scripts/pbuild/context.py
# Python build system
#
# Copyright (c) 2008, 2010 Zoltan Kovacs
# Copyright (c) 2010 Kornel Csernai
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of version 2 of the GNU General Public License
# as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import re
import os
import functions
import definitions
import logging

class ProjectContext :
    TOPLEVEL_FILE = "pbuild.toplevel"

    def __init__( self ) :
        self.toplevel = None

    def init( self ) :
        cwd = os.getcwd()
        if cwd[-1] == os.path.sep :
            cwd = cwd[:-1]

        while not os.path.isfile( cwd + os.path.sep + self.TOPLEVEL_FILE ) :
            if cwd == "":
                logging.error("Can't find toplevel.")
                return False

            index = cwd.rfind( os.path.sep )
            cwd = cwd[:index]

        self.toplevel = cwd

        return True

    def get_toplevel( self ) :
        return self.toplevel

class BuildContext :
    def __init__( self, pctx ) :
        self.pctx = pctx
        self.targets = []
        self.default_target = None
        self.definition_manager = definitions.DefinitionManager()
        self.function_handler = functions.FunctionHandler()

    def add_definition( self, definition ) :
        self.definition_manager.add_definition( definition )

    def remove_definition( self, name ) :
        self.definition_manager.remove_definition( name )

    def add_target( self, target ) :
        self.targets.append( target )

    def get_target( self, name, private_allowed = False ) :
        for target in self.targets :
            if target.get_name() == name and ( private_allowed or not target.is_private() ) :
                return target

        return None

    def get_definition( self, name ) :
        # handle special definitions first
        if name == "toplevel" :
            toplevel = self.pctx.get_toplevel()

            if toplevel == None :
                raise BaseException # todo
            else :
                return definitions.String( "", toplevel )

        # Try to find a user-defined stuff
        return self.definition_manager.get_definition( name )

    def get_definition_manager( self ) :
        return self.definition_manager

    def get_default_target( self ) :
        return self.default_target

    def get_project_context( self ) :
        return self.pctx

    def set_default_target( self, target ) :
        self.default_target = target

    def include( self, context ) :
        self.definition_manager.add_definitions(
            context.get_definition_manager().get_definitions()
        )

    def replace_definitions( self, text ) :
        while True :
            start = text.find( "${" )

            if start == -1 :
                break

            end = text.find( "}", start + 2 )

            if end == -1 :
                break;

            def_text = text[ start:end + 1 ]
            def_name = definitions.extract_definition_name( def_text )
            definition = self.get_definition( def_name )

            if definition == None :
                text = text[ :start ] + text[ end + 1: ]
            else :
                text = text[ :start ] + definition.to_string() + text[ end + 1: ]

        return text

    def replace_definitions_in_array( self, array ) :
        new_array = []

        for item in array :
            new_array.append( self.replace_definitions( item ) )

        return new_array

    def replace_wildcards( self, text ) :
        pos = text.find( "*" )

        if pos == -1 :
            return [ text ]

        # Wildcard allowed only in filenames, check it

        dir_pos = text.rfind( "/" )

        if dir_pos > pos :
            return [ text ]

        directory = text[ 0:dir_pos ]
        pattern = text[ dir_pos + 1: ]

        pattern = pattern.replace( ".", "\\." )
        pattern = pattern.replace( "*", ".*" )
        mask = re.compile( pattern )

        if not os.path.isdir( directory ) :
            return []

        entries = os.listdir( directory )

        result = []

        for entry in entries :
            if mask.match( entry ) :
                result.append( directory + os.sep + entry )

        return result

    def handle_everything( self, text ) :
        text = self.replace_definitions( text )
        return self.function_handler.handle_functions( text )

File: source/kernel/arch/i386/src/mm/context.c
/* Memory context handling code
 *
 * Copyright (c) 2009 Zoltan Kovacs
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <errno.h>
#include <kernel.h>
#include <console.h>
#include <mm/kmalloc.h>
#include <mm/context.h>
#include <mm/pages.h>
#include <lib/string.h>

#include <arch/mm/context.h>
#include <arch/mm/paging.h>

int arch_memory_context_init( memory_context_t* context ) {
    i386_memory_context_t* arch_context;

    /* Create the arch. specific memory context */

    arch_context = ( i386_memory_context_t* )kmalloc( sizeof( i386_memory_context_t ) );

    if ( arch_context == NULL ) {
        return -ENOMEM;
    }

    /* Allocate a memory page for the page directory */

    arch_context->page_directory = ( uint32_t* )alloc_pages( 1, MEM_COMMON );

    if ( arch_context->page_directory == NULL ) {
        kfree( arch_context );
        return -ENOMEM;
    }

    memsetl( arch_context->page_directory, 0, PAGE_SIZE / 4 );

    /* Set the arch. pointer in the memory context */

    context->arch_data = ( void* )arch_context;

    return 0;
}

int arch_memory_context_destroy( memory_context_t* context ) {
    int i;
    uint32_t* page_directory;
    i386_memory_context_t* arch_context;

    arch_context = ( i386_memory_context_t* )context->arch_data;

    if ( arch_context == NULL ) {
        return 0;
    }

    page_directory = arch_context->page_directory;

    for ( i = FIRST_USER_ADDRESS / PGDIR_SIZE; i < 1024; i++ ) {
        if ( page_directory[ i ] == 0 ) {
            continue;
        }

        free_pages( ( void* )( page_directory[ i ] & PAGE_MASK ), 1 );
    }

    free_pages( ( void* )page_directory, 1 );
    kfree( arch_context );

    return 0;
}

int arch_memory_context_clone( memory_context_t* old_context, memory_context_t* new_context ) {
    i386_memory_context_t* old_arch_context;
    i386_memory_context_t* new_arch_context;

    old_arch_context = ( i386_memory_context_t* )old_context->arch_data;
    new_arch_context = ( i386_memory_context_t* )new_context->arch_data;

    memcpy(
        new_arch_context->page_directory,
        old_arch_context->page_directory,
        FIRST_USER_ADDRESS / PGDIR_SIZE * sizeof( uint32_t )
    );

    return 0;
}

#ifdef ENABLE_DEBUGGER
int arch_memory_context_translate_address( memory_context_t* context, ptr_t linear, ptr_t* physical ) {
    uint32_t offset;
    uint32_t* pgd_entry;
    uint32_t* pt_entry;
    i386_memory_context_t* arch_context;

    offset = linear & ~PAGE_MASK;
    linear &= PAGE_MASK;

    arch_context = ( i386_memory_context_t* )context->arch_data;

    pgd_entry = &arch_context->page_directory[ PGD_INDEX( linear ) ];

    if ( ( ( *pgd_entry ) & PAGE_PRESENT ) == 0 ) {
        return -EINVAL;
    }

    pt_entry = &( ( uint32_t* )( *pgd_entry & PAGE_MASK ) )[ PT_INDEX( linear ) ];

    if ( ( ( *pt_entry ) & PAGE_PRESENT ) == 0 ) {
        return -EINVAL;
    }

    *physical = ( ( *pt_entry ) & PAGE_MASK ) + offset;

    return 0;
}
#endif /* ENABLE_DEBUGGER */
